### PR TITLE
ConfigureAwait (Mvc.Core)

### PR DESCRIPTION
**Pull request by benaadams**
Ref #3057 ConfigureAwait in Async methods; unless only one await and last
statement when return Task rather than awaiting

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ControllerActionExecutor.cs [hunk 1]
```diff
@@ -172,7 +172,7 @@ private static void ThrowIfWrappedTaskInstance(Type actualTypeReturned, string m
         /// </summary>
         private static async Task<object> CastToObject(Task task)
         {
-            await task;
+            await task.ConfigureAwait(false);
             return null;
```
Comment (benaadams) [inline src/Microsoft.AspNet.Mvc.Core/ControllerActionExecutor.cs]: This code is a bit exotic; was going to do something different; but probably needs a double check instead?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/JQueryFormValueProviderFactory.cs [hunk 2]
```diff
@@ -21,10 +21,12 @@ public class JQueryFormValueProviderFactory : IValueProviderFactory
 
             if (request.HasFormContentType)
             {
+                var culture = CultureInfo.CurrentCulture;
```
Comment (benaadams) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/JQueryFormValueProviderFactory.cs]: Want to use thread culture here? Request culture, as per https://github.com/aspnet/Mvc/pull/3059/files#diff-fbc3dd5ae10027009ad0c93d673e1141R19 may be more relevant?

File: src/Microsoft.AspNet.Mvc.Core/VirtualFileProviderResult.cs [hunk 3]
```diff
@@ -89,23 +89,19 @@ public string FileName
...
-
-                    return;
                 }
                 else
                 {
                     var fileStream = GetFileStream(fileInfo);
                     using (fileStream)
                     {
-                        await fileStream.CopyToAsync(response.Body, DefaultBufferSize, cancellation);
+                        return fileStream.CopyToAsync(response.Body, DefaultBufferSize, cancellation);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Core/VirtualFileProviderResult.cs]: This is broken. Can't return directly if a using is there.

File: src/Microsoft.AspNet.Mvc.Core/StreamOutputFormatter.cs [hunk 4]
```diff
@@ -40,7 +40,7 @@ public bool CanWriteResult([NotNull] OutputFormatterContext context, MediaTypeHe
                     response.ContentType = context.SelectedContentType.ToString();
                 }
 
-                await valueAsStream.CopyToAsync(response.Body);
+                return valueAsStream.CopyToAsync(response.Body);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Core/StreamOutputFormatter.cs]: This is broken. Can't return directly if a using is there.

File: src/Microsoft.AspNet.Mvc.Core/FileStreamResult.cs [hunk 5]
```diff
@@ -68,13 +68,13 @@ public Stream FileStream
...
         /// <inheritdoc />
-        protected async override Task WriteFileAsync(HttpResponse response, CancellationToken cancellation)
+        protected override Task WriteFileAsync(HttpResponse response, CancellationToken cancellation)
         {
             var outputStream = response.Body;
 
             using (FileStream)
             {
-                await FileStream.CopyToAsync(outputStream, BufferSize, cancellation);
+                return FileStream.CopyToAsync(outputStream, BufferSize, cancellation);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Core/FileStreamResult.cs]: This is broken. Can't return directly if a using is there.

File: src/Microsoft.AspNet.Mvc.Core/PhysicalFileProviderResult.cs [hunk 6]
```diff
@@ -86,10 +84,9 @@ public string FileName
 
                 using (fileStream)
                 {
-                    await fileStream.CopyToAsync(response.Body, DefaultBufferSize, cancellation);
+                    return fileStream.CopyToAsync(response.Body, DefaultBufferSize, cancellation);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Core/PhysicalFileProviderResult.cs]: This is broken. Can't return directly if a using is there.

File: src/Microsoft.AspNet.Mvc.Core/VirtualFileProviderResult.cs (hunk 3 above)
Comment (benaadams) [inline src/Microsoft.AspNet.Mvc.Core/VirtualFileProviderResult.cs]: Very good point; didn't think that through :(

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/FormValueProviderFactory.cs [hunk 7]
```diff
@@ -20,7 +20,7 @@ public class FormValueProviderFactory : IValueProviderFactory
 
                 return new ReadableStringCollectionValueProvider(
                     BindingSource.Form,
-                    await request.ReadFormAsync(),
+                    await request.ReadFormAsync().ConfigureAwait(false),
                     culture);
```
Comment (benaadams) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/FormValueProviderFactory.cs]: Culture call suggests is retrieved from request; but call gets it from current thread?

**Comment by rynowak** [comment]
@benaadams 

What's the measured perf impact of these changes?

**Comment by benaadams** [comment]
Need to wait till the myget feed updates

**Comment by PinpointTownes** [comment]
@rynowak related discussion: https://github.com/aspnet/Mvc/issues/3057

**Comment by benaadams** [comment]
Premise based on asp.net <= 4.5 but no longer applies to asp.net 5; will resubmit new PR with only asp.net 5 related changes.

**Pull request closed without merging**